<!doctype html>
<html>
<head>
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="chrome=1">
<title>CODOC</title>

<link rel="stylesheet" href="stylesheets/styles.css">
<link rel="stylesheet" href="stylesheets/pygment_trac.css">
<link rel="stylesheet" href="stylesheets/custom.css">
<meta name="viewport"
	content="width=device-width, initial-scale=1, user-scalable=no">
<!--[if lt IE 9]>
    <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
</head>
<body>
	<div class="wrapper">
		<header>
			<pre style="font-family: monospace;">
  _________  ___  ____  _____
 / ___/ __ \/ _ \/ __ \/ ___/
/ /__/ /_/ / // / /_/ / /__  
\___/\____/____/\____/\___/  
                            
        </pre>
			<p>CODOC is a compressed data format and API for coverage data
				stemming from sequencing experiments</p>

			<p class="view">
			<ul>
				<li><a href="https://github.com/popitsch/codoc/zipball/master">Download
						<strong>ZIP File</strong>
				</a></li>
				<li><a href="https://github.com/popitsch/codoc/tarball/master">Download
						<strong>TAR Ball</strong>
				</a></li>
				<li><a href="https://github.com/popitsch/codoc">View On <strong>GitHub</strong></a></li>
			</ul>
			</p>

			<p class="menu">
				<a href="#about">About CODOC</a>
			</p>
			<p class="menu">
				<a href="#about">Getting CODOC</a>
			</p>
			<p class="menu">
				<a href="#examples">Usage examples</a>
			</p>
			<p class="menu">
				<a href="#usage">Usage information</a>
			</p>
			<p class="menu">
				<a href="https://github.com/popitsch/codoc">View on GitHub</a>
			</p>

		</header>
		<section>
			<a name="about"></a>
			<h3>About</h3>

			<p>CODOC is an open file format and API for the lossless and
				lossy compression of depth-of-coverage (DOC) signals stemming from
				high-throughput sequencing (HTS) experiments. DOC data is a
				one-dimensional signal representing the number of reads covering
				each reference nucleotide in a HTS data set. DOC is highly-valuable
				information for interpreting (e.g., RNA-seq), filtering (e.g., SNP
				analysis) or detecting novel features (e.g., structural variants) in
				sequencing data sets.</p>
			<p>
				<img src="img/CoverageSignal.png" border="0" /> CODOC exploits
				several characteristics of DOC data stemming from HTS data and uses
				a non-uniform quantization model that preserves DOC signals in
				low-coverage regions nearly perfectly while allowing more divergence
				in highly covered regions.
			</p>
			<p>CODOC reduces required file sizes for DOC data up to 3500X
				when compared to raw representations and about 5-30X when compared
				to the other methods. The CODOC API supports efficient sequential
				and random access to compressed data sets.</p>
			</p>
			Usage scenarios for CODOC include (see below):
			<ul>
				<li>query coverage values at given genomic positions</li>
				<li>extract genomic regions with a given minimum/maximum
					coverage</li>
				<li>filter single-nucleotide variations in cancer/normal pairs</li>
				<li>linear combinations of coverage files (e.g., subtract one
					signal from another)</li>
				<li>cross-correlation of coverage signals</li>
			</ul>
			</p>

			<a name="building"></a>
			<h3>Getting CODOC</h3>

			<p>
				CODOC source code and <b>releases</b> can be downloaded from <a
					href="https://github.com/popitsch/codoc" target="_blank">GitHub</a>.
				CODOC uses <a href="http://maven.apache.org/" target="_blank">maven</a>
				as build tool. For development we recommend the <a
					href="http://www.eclipse.org/" target="_blank">Eclipse IDE for
					Java developers</a> and the <a href="http://www.eclipse.org/m2e/"
					target="_blank">m2e</a> Maven Integration for Eclipse.
			</p>
			<p>
				The CODOC jars can be built with 
				<code>bin/build-java.sh &lt;VERSION&gt;</code>
				(version is, e.g., <em>0.0.1</em>)
			</p>


			<a name="examples"></a></a>
			<h3>Usage Examples</h3>
			<p>
				Call the CODOC API w/o parameters to get short usage information (<a
					href="#usage">see below</a>) for the individual commands:
			<pre>
				<code>
$ java -jar bin/codoc-<em>0.0.1</em>.jar
$ java -jar bin/codoc-<em>0.0.1</em>.jar compress
$ java -jar bin/codoc-<em>0.0.1</em>.jar decompress
</code>
			</pre>

			The following command compressed coverage data extracted from a BAM
			file using a regular grid of height 5 for quantization. Only reads
			that have the SAM flag for PCR or optical duplicates (1024) unset and
			reads with a mapping quality > 20 are included.

			<pre>
				<code>
$ java -jar bin/codoc-0.0.1.jar compress -bam <em>BAM</em> -qmethod GRID -qparam 5 -filter "FLAGS^^1024" -filter "MAPQ>20" -o <em>BAM</em>.comp
</code>
			</pre>

			The following command prints the header of the compressed file
			showing the metadata as well as the used configuration for
			compression.

			<pre>
				<code>
$ java -jar bin/codoc-0.0.1.jar decompress head -covFile <em>BAM</em>.comp
</code>
			</pre>

			The following command starts an interactive query session for the
			compressed file.

			<pre>
				<code>
$ java -jar bin/codoc-0.0.1.jar decompress query -covFile <em>BAM</em>.comp
</code>
			</pre>

			The following command extracts all regions from the compressed file
			that have a coverage between MIN and MAX and stores them to a <a
				href="http://genome.ucsc.edu/FAQ/FAQformat.html#format1"
				target="_blank">BED</a> file. The given name/description are used
			for the BED track.
			<pre>
				<code>
$ java -jar bin/codoc-0.0.1.jar decompress tobed -covFile <em>BAM</em>.comp -outFile <em>BED</em> -min <em>MIN</em> -max <em>MAX</em> -name <em>NAME</em> -description <em>DESCRIPTION</em>
</code>
			</pre>


			NOTE: if you run into OutOfMemory errors, you can increase the
			available heap space of the java JVM (e.g., to 4 GByte) using the
			following switch:
			<pre>java -Xmx4g ...</pre>


			</p>

			<a name="usage"></a>
			<h3>Usage</h3>
			<p>Usage of the compressor:
			<pre>
				<code>
Usage:  java -jar x.jar compress -bam &lt;arg&gt; | -cov &lt;arg&gt; [-bedFile &lt;arg&gt;] [-blockSize &lt;arg&gt;] [-compressionAlgorithm
       &lt;arg&gt;]  [-createStats] [-dumpRawCoverage] [-filter &lt;arg&gt;] [-keepWorkDir] -o &lt;arg&gt; [-qmethod &lt;arg&gt;] [-qparam
       &lt;arg&gt;] [-scaleCoverage &lt;arg&gt;] [-statsFile &lt;arg&gt;] [-t &lt;arg&gt;] [-v] [-vcf &lt;arg&gt;]

Mandatory Params
        -bam,--bamFile &lt;arg&gt;  Input BAM file (using this will exclude the -cov parameter)
        -cov,--covFile &lt;arg&gt;  Input coverage file (using this will exclude the -bam parameter)
        -o,--out &lt;arg&gt;        Output file

Params
        -bam,--bamFile &lt;arg&gt;         Input BAM file (using this will exclude the -cov parameter)
        -bedFile &lt;arg&gt;               Input BED file containing regions of interest (optional). The file will be
                                     interpreted 0-based!
        -blockSize &lt;arg&gt;             Number of codewords per data block (default: 1000000.0)
        -compressionAlgorithm &lt;arg&gt;  Used envelope compression algorithm [NONE, GZIP, BZIP2]. Default method is GZIP
        -cov,--covFile &lt;arg&gt;         Input coverage file (using this will exclude the -bam parameter)
        -createStats                 Create a statistics file (default: false)
        -dumpRawCoverage             Create raw coverage file (&lt;outfile&gt;.rawcoverage). No heading or trailing 0-coverage
                                     zones will be included. Output will be 1-based coordinates.
        -filter &lt;arg&gt;                Filter used reads by SAM attribute (applicable only in combination with -bam, not
                                     -cov). Multiple filters can be provided that will be combined by a logical AND.
                                     Note that filters have no effect on reads that have no value for the according
                                     attribute. Examples: 'X0&gt;9', 'X1=ABC', 'FLAGS^^512', 'none',... [default: -filter
                                     FLAGS^^1024 -filter FLAGS^^512].See below for more help on filters.
        -keepWorkDir                 Do not delete the work directory (e.g., for debugging purposes)
        -o,--out &lt;arg&gt;               Output file
        -qmethod &lt;arg&gt;               Quantization method (PERC, GRID, LOG2); default: PERC
        -qparam &lt;arg&gt;                Quantization parameter (default: 0.2)
        -scaleCoverage &lt;arg&gt;         Optional parameter for scaling the coverage values. Applicable only for coverage
                                     file input (not for BAM files). (default: none)
        -statsFile &lt;arg&gt;             Statistics file name (default: &lt;outfile&gt;.stats)
        -t,--temp &lt;arg&gt;              Temporary working directory (default is current dir)
        -v,--verbose                 be verbose
        -vcf,--vcfFile &lt;arg&gt;         Input VCF file (optional)


        -------------------------------- FILTERS -----------------------------------
        Filters can be used to restrict the coverage extraction from am SAM/BAM file
        to reads that match the given filter criteria. Multiple filters are combined
        by a logical AND. Filters are of the form &lt;FIELD&gt;&lt;OPERATOR&gt;&lt;VALUE&gt;

        Possible fields:
        MAPQ    the mapping quality
        FLAGS   the read flags
        Other names will be mapped directly to the optional field name in the SAM file.
        Use e.g., NM for the 'number of mismatches' field. Reads that do not have a field
        set will be included. @see http://samtools.sourceforge.net/SAMv1.pdf

        Possible operators:
        &lt;, &lt;=, =, &gt;, &gt;=, ^^ (flag unset), &amp;&amp; (flag set)

        Example: (do NOT use reads with mapping quality &lt;= 20, or multiple perfect hits)
        -filter 'MAPQ&gt;20' -filter 'H0=1'
        ----------------------------------------------------------------------------
</code>
			</pre>


			Usage of the decompressor HEAD command:
			<pre>
				<code>
Usage:    java -jar x.jar decompress HEAD -covFile &lt;arg&gt; [-keepWorkDir] [-tmpDir &lt;arg&gt;] [-v]
Params:
          -covFile,--cov &lt;arg&gt;  Compressed coverage input file.
          -keepWorkDir          Do not delete the work directory (e.g., for debugging purposes).
          -tmpDir,--temp &lt;arg&gt;  Temporary working directory (default is current dir).
          -v,--verbose          be verbose.

Prints the header from a compressed coverage file.
</code>
			</pre>

			Usage of the decompressor QUERY command:
			<pre>
				<code>
Usage:    java -jar x.jar decompress QUERY [-chrLenFile &lt;arg&gt;] -covFile &lt;arg&gt; [-keepWorkDir] [-tmpDir &lt;arg&gt;] [-v] [-vcfFile &lt;arg&gt;]
Params:
          -chrLenFile &lt;arg&gt;     Chromosome lengths (file format: chr\tlength\n). Use to get proper padding for zero-coverage regions.
          -covFile,--cov &lt;arg&gt;  Compressed coverage input file.
          -keepWorkDir          Do not delete the work directory (e.g., for debugging purposes).
          -tmpDir,--temp &lt;arg&gt;  Temporary working directory (default is current dir).
          -v,--verbose          be verbose.
          -vcfFile &lt;arg&gt;        VCF file used for the compression (optional).

Starts an interactive query session.
</code>
			</pre>

			Usage of the decompressor TOBED command:
			<pre>
				<code>
Usage:    java -jar x.jar decompress TOBED [-additional &lt;arg&gt;] [-chrLenFile &lt;arg&gt;] -covFile &lt;arg&gt; [-description &lt;arg&gt;] [-keepWorkDir] [-max &lt;arg&gt;] [-min &lt;arg&gt;]
       [-name &lt;arg&gt;] -outFile &lt;arg&gt; [-tmpDir &lt;arg&gt;] [-v] [-vcfFile &lt;arg&gt;]
Params:
          -additional &lt;arg&gt;     Additional BED track info (optional).
          -chrLenFile &lt;arg&gt;     Chromosome lengths (file format: chr\tlength\n). Use to get proper padding for zero-coverage regions.
          -covFile,--cov &lt;arg&gt;  Compressed coverage input file.
          -description &lt;arg&gt;    BED description (optional).
          -keepWorkDir          Do not delete the work directory (e.g., for debugging purposes).
          -max &lt;arg&gt;            maximum coverage (omit to leave unrestricted).
          -min &lt;arg&gt;            minimum coverage (omit to leave unrestricted).
          -name &lt;arg&gt;           BED track name (optional).
          -outFile &lt;arg&gt;        Output file that will contain the BED intervals.
          -tmpDir,--temp &lt;arg&gt;  Temporary working directory (default is current dir).
          -v,--verbose          be verbose.
          -vcfFile &lt;arg&gt;        VCF file used for the compression (optional).

Creates a BED file containing regions that are above/below given coverage thresholds.
</code>
			</pre>

			Usage of the decompressor TOWIG command:
			<pre>
				<code>
Usage:    java -jar x.jar decompress TOWIG [-chrLenFile &lt;arg&gt;] -covFile &lt;arg&gt; [-keepWorkDir] [-tmpDir &lt;arg&gt;] [-v] [-vcfFile &lt;arg&gt;] -wigFile &lt;arg&gt;
Params:
          -chrLenFile &lt;arg&gt;     Chromosome lengths (file format: chr\tlength\n). Use to get proper padding for zero-coverage regions.
          -covFile,--cov &lt;arg&gt;  Compressed coverage input file.
          -keepWorkDir          Do not delete the work directory (e.g., for debugging purposes).
          -tmpDir,--temp &lt;arg&gt;  Temporary working directory (default is current dir).
          -v,--verbose          be verbose.
          -vcfFile &lt;arg&gt;        VCF file used for the compression (optional).
          -wigFile &lt;arg&gt;        Output file that will contain the rounded approximated coverage values (optional).

Converts the coverage data to a WIG file.
</code>
			</pre>



			</p>

		</section>

		<footer>
			<p>
				<small>This project is maintained by <a
					href="https://github.com/popitsch">niko popitsch</a></small>
			</p>
			<!--  <p><small>Hosted on GitHub Pages &mdash; Theme by <a href="https://github.com/orderedlist">orderedlist</a></small></p> -->
		</footer>
	</div>
	<script src="javascripts/scale.fix.js"></script>

</body>
</html>